quant bn/gn/in: move scale and zp to buffers

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45313 quant bn/gn/in: move scale and zp to buffers**

Summary:

Fixes https://github.com/pytorch/pytorch/issues/43774

Currently the scale and zero_point in these modules is not
saved in the state_dict, so saving them and loading them
back does not work.  Fixing by making them buffers.

Note: not handling loading old state_dicts on purpose as unfortunately
there isn't a way to recover correct scales and zero_points as we did
not save them.

The # of people affected by this should be small,
as most people fuse BNs into preceding layers during quantization, and quantized
BN/IN support is very new.

Test Plan:

```
python test/test_quantization.py TestStaticQuantizedModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23919143](https://our.internmc.facebook.com/intern/diff/D23919143)

vkuzo committed:
quant bn/gn/in: move scale and zp to buffers

Summary:

Fixes https://github.com/pytorch/pytorch/issues/43774

Currently the scale and zero_point in these modules is not
saved in the state_dict, so saving them and loading them
back does not work.  Fixing by making them buffers.

Note: not handling loading old state_dicts on purpose as unfortunately
there isn't a way to recover correct scales and zero_points as we did
not save them.

The # of people affected by this should be small,
as most people fuse BNs into preceding layers during quantization, and quantized
BN/IN support is very new.

Test Plan:

```
python test/test_quantization.py TestStaticQuantizedModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/nn/quantized/modules/normalization.py:
```diff
@@ -120,13 +121,13 @@ def __init__(self, num_features, weight, bias, scale, zero_point,
         self.bias = bias
-        self.scale = scale
-        self.zero_point = zero_point
+        self.register_buffer('scale', torch.Tensor(scale))
```
make scale float?

jerryzh168 reviewed (approved):
please make scale Tensor float as well

vkuzo commented:
> please make scale Tensor float as well

sure, it defaults to float but will add it to clarify intent

vkuzo committed:
Update on "quant bn/gn/in: move scale and zp to buffers"


Summary:

Fixes https://github.com/pytorch/pytorch/issues/43774

Currently the scale and zero_point in these modules is not
saved in the state_dict, so saving them and loading them
back does not work.  Fixing by making them buffers.

Note: not handling loading old state_dicts on purpose as unfortunately
there isn't a way to recover correct scales and zero_points as we did
not save them.

The # of people affected by this should be small,
as most people fuse BNs into preceding layers during quantization, and quantized
BN/IN support is very new.

Test Plan:

```
python test/test_quantization.py TestStaticQuantizedModule
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D23919143](https://our.internmc.facebook.com/intern/diff/D23919143)

[ghstack-poisoned]

z-a-f reviewed (approved):
lgtm

vkuzo commented:
abandoning, doesn't seem worth it yet to fix all the callsites depending on the existing behavior, we may revisit in the future
